### PR TITLE
[ARCTIC_1726]flink1.12 When double write into fliestore and logstore, AutomaticLogWriter skip doing snapshotState, Leading to NoSuchElementException during initializeState after failover and restart 

### PR DESCRIPTION
**Pull request by xieyi888**
issue: https://github.com/NetEase/arctic/issues/1726

## Why are the changes needed?

For Flink1.12, When auto double write into fliestore and logstore, AutomaticLogWriter skip doing snapshotState, Leading to NoSuchElementException during initializeState after failover and restart because checkpointedState is empty

## Brief change log
- for Flink 1.12,Flink1.14,Flink 1.15, remove 'arctic.emit.auto-write-to-logstore.watermark-gap' configuration, after watermark has catched up the watermark-gap. In order to double write immediately after job failover restart
- for Flink 1.12,  add isValidCheckpointState method in com.netease.arctic.flink.write.hidden.AbstractHiddenLogWriter,to avoid  empty checkpointedState and NoSuchElementException after job  failover restart 

## How was this patch tested?
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] Run test locally before making a pull request

## Documentation

  - Does this pull request introduces a new feature? (yes / no)
  no
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by xieyi888** [comment]
Sorry, It seems has a logical error, once a flink job( auto double write)  watermark catched up the 'log-store.catch-up-timestamp',  
table Properties 'log-store.catch-up' would be set true Persistently
https://github.com/NetEase/arctic/blob/master/flink/v1.12/flink/src/main/java/com/netease/arctic/flink/write/AutomaticDoubleWriteStatus.java#L76

Another new flink job( auto double write)，even without watermark,  would  double write immediately  once the job is launched.

**Comment by lklhdu** [comment]
> Sorry, Maybe It seems has a unreasonable logical, once a flink job( auto double write) watermark catched up the 'log-store.catch-up-timestamp', table Properties 'log-store.catch-up' would be set true Persistently
> 
> ```
>   public void processWatermark(Watermark mark) {
>     if (isDoubleWrite()) {
>       return;
>     }
>     if (specification.shouldDoubleWrite(mark.getTimestamp())) {
>       shouldDoubleWrite = true;
>       LOG.info("processWatermark {}, subTaskId is {}, should double write is true.", mark, subtaskId);
>       LOG.info("begin update arctic table, set {} to true", LOG_STORE_CATCH_UP.key());
>       UpdateProperties updateProperties = table.updateProperties();
>       updateProperties.set(LOG_STORE_CATCH_UP.key(), String.valueOf(true));
>       updateProperties.set(LOG_STORE_CATCH_UP_TIMESTAMP.key(), String.valueOf(System.currentTimeMillis()));
>       updateProperties.commit();
>       LOG.info("end update arctic table.");
>     }
>   }
> 
>   public void sync() {
>     table.refresh();
>     Map<String, String> properties = table.properties();
>     shouldDoubleWrite =
>         Boolean.parseBoolean(
>             properties.getOrDefault(LOG_STORE_CATCH_UP.key(), String.valueOf(LOG_STORE_CATCH_UP.defaultValue())));
>     LOG.info("AutomaticDoubleWriteStatus sync, subTaskId: {}, should double write: {}", subtaskId, shouldDoubleWrite);
>   }
> ```
> 
> Another new flink job( auto double write)，even without define watermark, would double write immediately once the job is launched.
> 
> Maybe we should try some better solutions

We can decide whether or not to double-write by determining whether or not the task has the 'arctic.emit.auto-write-to-logstore.watermark-gap' configuration set: if it is configured, it will not write to the logstore until the watermark of arctic writers is greater than the current system timestamp minus the configured watermark-gap; if it is not configured, then it will write directly to the logstore.

**Comment by lklhdu** [comment]
Can you please review this pr? @YesOrNo828

**Review by YesOrNo828 — COMMENTED** [review]
@xieyi888  Thanks for your contribution. Could you update the guide about the `'arctic.emit.mode'='auto'` configuration.

File: flink/v1.12/flink/src/main/java/com/netease/arctic/flink/write/hidden/AbstractHiddenLogWriter.java [hunk 1]
```diff
@@ -139,7 +139,7 @@ public void initializeState(StateInitializationContext context) throws Exception
                 helper));
     int parallelism = getRuntimeContext().getNumberOfParallelSubtasks();
 
-    if (context.isRestored() && parallelismSame(parallelism)) {
+    if (context.isRestored() && parallelismSame(parallelism) && isValidCheckpointState()) {
```
Comment (YesOrNo828) [inline flink/v1.12/flink/src/main/java/com/netease/arctic/flink/write/hidden/AbstractHiddenLogWriter.java:142]: Does Flink 1.14/1.15 need to be modified as well? I don't see any changes in this PR

**Comment by lklhdu** [comment]
This pr will also reslove #1036 #1601

**Review by YesOrNo828 — APPROVED** [review]
LGTM

**Pull request merged**